applications/ShapeOptimizationApplication/custom_utilities/tree_based_functions.h

```diff
@@ -0,0 +1,299 @@
+// ==============================================================================
+//  KratosShapeOptimizationApplication
+//
+//  License:         BSD License
+//                   license: ShapeOptimizationApplication/license.txt
+//
+//  Main authors:    Baumgaertner Daniel, https://github.com/dbaumgaertner
+//
+// ==============================================================================
+
+#ifndef TREE_BASED_FUNCTIONS
+#define TREE_BASED_FUNCTIONS
+
+// ------------------------------------------------------------------------------
+// System includes
+// ------------------------------------------------------------------------------
+#include <iostream>
+#include <string>
+#include <algorithm>
+#include <unordered_map>
+
+// ------------------------------------------------------------------------------
+// Project includes
+// ------------------------------------------------------------------------------
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "spatial_containers/spatial_containers.h"
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
+/** Detail class definition.
+
+*/
+
+class TreeBasedFunctions
```

Comment:
Maybe `NodeSearchBasedFunctions` is more clear to a "user". The fact that its tree based node search is internal information.

Reply:
NodeSearchBasedFunctions sounds a bit clunky. What about SearchBasedFunctions?